FirebaseDatabase/Sources/Core/FRepoInfo.m

```diff
@@ -25,61 +25,94 @@ @interface FRepoInfo ()
 
-@synthesize namespace;
-@synthesize host;
-@synthesize internalHost;
-@synthesize secure;
-@synthesize domain;
-
-- (id)initWithHost:(NSString *)aHost
-          isSecure:(bool)isSecure
-     withNamespace:(NSString *)aNamespace {
+- (instancetype)init {
+    [NSException
+         raise:@"FIRDatabaseInvalidInitializer"
+        format:@"Invalid initializer invoked. This is probably a bug in RTDB."];
+    abort();
+}
+
+- (instancetype)initWithHost:(NSString *)aHost
+                    isSecure:(BOOL)isSecure
+               withNamespace:(NSString *)aNamespace
+              underlyingHost:(NSString *)underlyingHost {
     self = [super init];
     if (self) {
-        host = aHost;
-        domain =
-            [host containsString:@"."]
-                ? [host
-                      substringFromIndex:[host rangeOfString:@"."].location + 1]
-                : host;
-        secure = isSecure;
-        namespace = aNamespace;
+        _underlyingHost = [underlyingHost copy];
+        _host = [aHost copy];
+        _domain =
+            [_host containsString:@"."]
+                ? [_host
+                      substringFromIndex:[_host rangeOfString:@"."].location +
+                                         1]
+                : _host;
+        _secure = isSecure;
+        _namespace = aNamespace;
 
         // Get cached internal host if it exists
         NSString *internalHostKey =
-            [NSString stringWithFormat:@"firebase:host:%@", self.host];
+            [NSString stringWithFormat:@"firebase:host:%@", _host];
         NSString *cachedInternalHost = [[NSUserDefaults standardUserDefaults]
             stringForKey:internalHostKey];
         if (cachedInternalHost != nil) {
-            internalHost = cachedInternalHost;
+            _internalHost = cachedInternalHost;
         } else {
-            internalHost = self.host;
+            _internalHost = [_host copy];
         }
     }
     return self;
 }
 
+- (instancetype)initWithHost:(NSString *)host
+                    isSecure:(BOOL)secure
+               withNamespace:(NSString *)namespace {
+    return [self initWithHost:host
+                     isSecure:secure
+                withNamespace:namespace
+               underlyingHost:nil];
+}
+
+- (instancetype)initWithInfo:(FRepoInfo *)info emulatedHost:(NSString *)host {
+    self = [self initWithHost:host
+                     isSecure:info.secure
+                withNamespace:info.namespace
+               underlyingHost:info.host];
+
+    // Compute the domain based on the original host instead of the emulated
+    // host in an emulated setting.
+    if (self != nil) {
+        _domain =
+            [_underlyingHost containsString:@"."]
+                ? [_underlyingHost
+                      substringFromIndex:[_underlyingHost rangeOfString:@"."]
+                                             .location +
+                                         1]
+                : _underlyingHost;
```

Comment:
This also seems missing in the other ports. Does `localhost` now work for `_domain`?

Reply:
I did this because the existing implementation of `isCustomHost` depended on `domain`, which was derived from `host`. `localhost` works fine for `domain`, but setting that in the initializer makes the database instance consider the emulator to be a custom domain.

Reply:
After talking to @yuchenshi, I think we want the SDK to consider the emulator URLs as a custom domain. We essentially want to change all internal URLs to "localhost" but still allow production URLs in "refFromUrl". We can talk about this more offline (and still change our plan if need be).